File: sdks/python/gbz185_sdk/transport.py
from __future__ import annotations

from abc import ABC, abstractmethod
import json
from typing import Any, Callable
from urllib import request

JsonValue = Any
JsonOperationHandler = Callable[[JsonValue], JsonValue]


class JsonTransport(ABC):
    @abstractmethod
    def request(self, operation: str, payload: JsonValue) -> JsonValue:
        raise NotImplementedError


class InProcessJsonTransport(JsonTransport):
    def __init__(self) -> None:
        self._handlers: dict[str, JsonOperationHandler] = {}

    def register(self, operation: str, handler: JsonOperationHandler) -> None:
        self._handlers[operation] = handler

    def request(self, operation: str, payload: JsonValue) -> JsonValue:
        if operation not in self._handlers:
            raise KeyError(f"No JSON transport handler registered for operation: {operation}")
        return self._handlers[operation](payload)


class HttpJsonTransport(JsonTransport):
    def __init__(self, endpoint: str, headers: dict[str, str] | None = None, timeout: float = 30.0) -> None:
        self.endpoint = endpoint
        self.headers = headers or {}
        self.timeout = timeout

    def request(self, operation: str, payload: JsonValue) -> JsonValue:
        body = json.dumps({"operation": operation, "payload": payload}).encode("utf-8")
        headers = {"content-type": "application/json", **self.headers}
        req = request.Request(self.endpoint, data=body, headers=headers, method="POST")
        with request.urlopen(req, timeout=self.timeout) as response:
            return json.loads(response.read().decode("utf-8"))

File: src/transport.ts
export interface JsonTransport {
  request<TRequest = unknown, TResponse = unknown>(operation: string, payload: TRequest): Promise<TResponse>;
}

export type JsonOperationHandler<TRequest = unknown, TResponse = unknown> = (payload: TRequest) => Promise<TResponse> | TResponse;

export class InProcessJsonTransport implements JsonTransport {
  private handlers = new Map<string, JsonOperationHandler>();

  register<TRequest = unknown, TResponse = unknown>(operation: string, handler: JsonOperationHandler<TRequest, TResponse>): void {
    this.handlers.set(operation, handler as JsonOperationHandler);
  }

  async request<TRequest = unknown, TResponse = unknown>(operation: string, payload: TRequest): Promise<TResponse> {
    const handler = this.handlers.get(operation);
    if (!handler) {
      throw new Error(`No JSON transport handler registered for operation: ${operation}`);
    }
    return handler(payload) as Promise<TResponse>;
  }
}

export interface HttpJsonTransportOptions {
  endpoint: string;
  fetchImpl?: typeof fetch;
  headers?: Record<string, string>;
}

export class HttpJsonTransport implements JsonTransport {
  private readonly fetchImpl: typeof fetch;

  constructor(private readonly options: HttpJsonTransportOptions) {
    this.fetchImpl = options.fetchImpl ?? fetch;
  }

  async request<TRequest = unknown, TResponse = unknown>(operation: string, payload: TRequest): Promise<TResponse> {
    const response = await this.fetchImpl(this.options.endpoint, {
      method: "POST",
      headers: {
        "content-type": "application/json",
        ...(this.options.headers ?? {})
      },
      body: JSON.stringify({ operation, payload })
    });
    if (!response.ok) {
      throw new Error(`HTTP JSON transport failed: ${response.status} ${response.statusText}`);
    }
    return (await response.json()) as TResponse;
  }
}
